Closes #222

JohnGiorgi opened the pull request:
- **Name:** *PubTator Central*
- **Description:** *PubTator Central (PTC, https://www.ncbi.nlm.nih.gov/research/pubtator/) is a web service for
exploring and retrieving bioconcept annotations in full text biomedical articles. PTC provides
automated annotations from state-of-the-art text mining systems for genes/proteins, genetic
variants, diseases, chemicals, species and cell lines, all available for immediate download.*
- **Paper:** https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3692066/
- **Data:** https://ftp.ncbi.nlm.nih.gov/pub/lu/PubTatorCentral/

### Checkbox

- [X] Confirm that this PR is linked to the dataset issue.
- [X] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [X] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [X] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [X] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.

Closes #222.

JohnGiorgi commented:
Okay this is mostly done, just wanted to note some stuff as well as some outstanding peculiarities about the dataset:

__Notes__:

- I followed the `chemprot.py` schema pretty closely and modified it to be as close to the PubTator Central format as possible. This led to some decisions like breaking “text” into “title” and “abstract”, renaming “entities” to “mentions” and “id” to “concept_id”.
- I wrote `_parse_pubtator_file` and `_pubtator_parse_to_bigbio_kb` as analogs to `parsing.parse_brat_file` and `parsing.brat_parse_to_bigbio_kb`. It may be useful to move them to `parsing` if there is at least one other dataset in the PubTator format (I believe BC5CDR is but it also comes as BioC).

__Outstanding__:

- There are no splits, so I just wrote one `SplitGenerator` with the `name=datasets.Split.TRAIN`
- This dataset is updated monthly, according to its README.txt. So for `_SOURCE_VERSION`, I set it to whatever the current timestamp is on the [FTP server](https://ftp.ncbi.nlm.nih.gov/pub/lu/PubTatorCentral/). There's only one URL though, so not sure how `_SOURCE_VERSION` would be kept up to date?
- I am not sure what the `_LICENSE` should be. The text is the title and abstracts of PubMed and PMC papers and the annotations are from machine learning systems. I would think public domain?

galtay commented:
hi @JohnGiorgi great work so far! can I ask that you implement this using the files that end in `sample` from the FTP for now so that we can try some quick local testing?

JohnGiorgi commented:
hi @galtay, the `.sample` files appear to be in a different format than the `.offset`, so I don't think one can be exchanged for the other. Most importantly they don't contain any offsets for the entities, so they wouldn't be suitable for NER.

Here are the first couple of lines of [bioconcepts2pubtatorcentral.sample](https://ftp.ncbi.nlm.nih.gov/pub/lu/PubTatorCentral/bioconcepts2pubtatorcentral.sample)

```
3958000	Species	9606	patients	SR4GN
3958000	Disease	MESH:D012544	Scheuermann's kyphosis	TaggerOne
3958000	Disease	MESH:D011183		MESH
3958000	Disease	MESH:D007738		MESH
```

galtay commented:
haha, there's a sample file for everything except the thing we need :) ... ok, how about this ... can we maybe implement a config that uses the API to grab a small number of PMIDs? I think you can do that like this, 
```
import requests
blob = {
    "pmids": [
        "27940449",
        "28058064",
        "28078498"
    ],
    "concepts": [
        "gene",
        "disease"
    ]
}
url = "https://www.ncbi.nlm.nih.gov/research/pubtator-api/publications/export/pubtator"
res = requests.post(url, json=blob)                                                                                  
content = res.content.decode("utf-8") 
```

I think its great to have a big dataset like this in the repo, but I also think it would be great to have a way test on a small subset. The above is just an example, we could make this (say) 100 PMIDs and all the concepts if we like.

JohnGiorgi commented:
Okay cool, I have implemented something that works. You can see my latest commit but the highlights are:

- Add new config `"pubtator_central_api"` with a new schema "`api`" (although in truth the data format is the same as the `"source"` schema).
- If `self.config.schema` is `"api"` it makes a post request to the api. The PMIDs are hardcoded for now, not sure they best way to pass them in?
- Everything else was mostly unchanged.

I can now call

```python
>>> dataset = load_dataset("biodatasets/pubtator_central/pubtator_central.py", name="pubtator_central_source")
>>> dataset["train"][0]
>>> {'pmid': '27940449',
 'title': 'Next Generation Sequencing of Circulating Cell-Free DNA for Evaluating Mutations and Gene Amplification in Metastatic Breast Cancer.',
 'abstract': 'BACKGROUND: Breast cancer tissues are heterogeneous and show diverse somatic mutations and somatic copy number alterations (CNAs). We used a novel targeted next generation sequencing (NGS) panel to examine cell-free DNA (cfDNA) to detect somatic mutations and gene amplification in women with metastatic breast cancer (MBC). METHODS: cfDNA from pretreated patients (n = 42) and 9 healthy controls were compared with matched lymphocyte DNA by NGS, using a custom 158 amplicon panel covering hot-spot mutations and CNAs in 16 genes, with further validation of results by droplet digital PCR. RESULTS: No mutations were identified in cfDNA of healthy controls, whereas exactly half the patients with metastatic breast cancer had at least one mutation or amplification in cfDNA (mean 2, range 1-6) across a total of 13 genes. Longitudinal follow up showed dynamic changes to mutations and gene amplification in cfDNA indicating clonal and subclonal response to treatment that was more dynamic than cancer antigen 15-3 (CA15-3). Interestingly, at the time of blood sampling disease progression was occurring in 7 patients with erb-b2 receptor tyrosine kinase 2 (ERBB2) gene amplification in their cfDNA and 3 of these patients were human epidermal growth factor receptor 2 (HER2) negative at diagnosis, suggesting clonal evolution to a more aggressive phenotype. Lastly, 6 patients harbored estrogen receptor 1 (ESR1) mutations in cfDNA, suggesting resistance to endocrine therapy. Overall 9 of 42 patients (21%) had alterations in cfDNA that could herald a change in treatment. CONCLUSIONS: Targeted NGS of cfDNA has potential for monitoring response to targeted therapies through both mutations and gene amplification, for analysis of dynamic tumor heterogeneity and stratification to targeted therapy.',
 'mentions': [{'concept_id': 'MESH:D001943',
   'type': 'Disease',
   'text': 'Metastatic Breast Cancer',
   'offsets': [107, 131]},
  {'concept_id': 'MESH:D001943',
   'type': 'Disease',
   'text': 'Breast cancer',
   'offsets': [145, 158]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'women',
   'offsets': [415, 420]},
  {'concept_id': 'MESH:D001943',
   'type': 'Disease',
   'text': 'breast cancer',
   'offsets': [437, 450]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [489, 497]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [816, 824]},
  {'concept_id': 'MESH:D001943',
   'type': 'Disease',
   'text': 'breast cancer',
   'offsets': [841, 854]},
  {'concept_id': '4582',
   'type': 'Gene',
   'text': 'cancer antigen 15-3',
   'offsets': [1127, 1146]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [1241, 1249]},
  {'concept_id': '2064',
   'type': 'Gene',
   'text': 'erb-b2',
   'offsets': [1255, 1261]},
  {'concept_id': '2064',
   'type': 'Gene',
   'text': 'ERBB2',
   'offsets': [1290, 1295]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [1346, 1354]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'human',
   'offsets': [1360, 1365]},
  {'concept_id': '2064',
   'type': 'Gene',
   'text': 'epidermal growth factor receptor 2',
   'offsets': [1366, 1400]},
  {'concept_id': '2064',
   'type': 'Gene',
   'text': 'HER2',
   'offsets': [1402, 1406]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [1501, 1509]},
  {'concept_id': '2099',
   'type': 'Gene',
   'text': 'estrogen receptor 1',
   'offsets': [1519, 1538]},
  {'concept_id': '2099',
   'type': 'Gene',
   'text': 'ESR1',
   'offsets': [1540, 1544]},
  {'concept_id': '9606',
   'type': 'Species',
   'text': 'patients',
   'offsets': [1626, 1634]},
  {'concept_id': 'MESH:D009369',
   'type': 'Disease',
   'text': 'tumor',
   'offsets': [1873, 1878]}]}

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,317 @@
+        "ProteinMutation": None,
+        "DNAMutation": None,
+        "DomainMotif": None,
+    }
```
do these db names come from the data itself? 
is it worthwhile to put urls in this module somewhere (e.g. https://www.ncbi.nlm.nih.gov/gene/) ?

do you get anything interesting when googling some of the IDs that come back for the 3 that you have `None` for above? sometimes you get lucky ... e.g. googling "rs113488022" will lead you to ncbi_dbsnp

JohnGiorgi replied:
I don't believe so, but they do come from PubTator, see an example below:

```
```

I agree that URLs would be better. Is there a certain key/field of the `bigbio_kb` schema that I could use? Or should we consider adding a key, e.g. `db_url`

galtay replied:
good question ... I was thinking more of a module level constant for the URLs ... hopefully at the end of the hackathon we can produce something like a mapping between every db_name and a URL in some utils module.

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,317 @@
+
+    def _info(self) -> datasets.DatasetInfo:
+
+        if self.config.schema in ["api", "source"]:
```
with changes above, this could just be `self.config.schema == "source"`

galtay reviewed (changes requested):
A few suggestions on how to organize configs. Also looks like some unrelated files got pulled into this PR.

JohnGiorgi commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,317 @@
+# Note that this doesn't have to be a common open source license.
+# Some datasets have custom licenses. In this case, simply put the full license terms
+# into `_LICENSE`
+_LICENSE = ""
```
I looked around their website and papers but can't find a specific licence. I have emailed the PubTator maintainers, just waiting on a response!

JohnGiorgi replied:
Okay, I have added the license the maintainers sent me. It is a custom licence about a paragraph in length.

JohnGiorgi commented:
> A few suggestions on how to organize configs. Also looks like some unrelated files got pulled into this PR.

Ah, sorry about that. I was trying to keep the branch up-to-date with master by following the instructions:

```
git fetch upstream
git pull
```

I don't know why GitHub is showing these as new commits?

hakunanatasha commented:
@JohnGiorgi the readme is probably from the progress bar updating; the rest possibly from hotfixes.

Thank you for keeping up to date with the main branch :P

galtay commented:
> No worries :) Should I leave it as is or is there some kind of git incantation I can use so that these commits no longer appear under this PR?

you can probably follow Case: 3 here https://ffeathers.wordpress.com/2020/03/21/removing-an-updated-file-from-a-pr-on-github/

galtay commented:
fyi bioc update is here now https://github.com/bigscience-workshop/biomedical/pull/397

JohnGiorgi commented:
> hi @JohnGiorgi great work so far! can I ask that you implement this using the files that end in `sample` from the FTP for now so that we can try some quick local testing?

Hi @galtay, sorry I hadn't forgot about this PR! Just focused on working on two other data loaders first. When you originally suggested the above there was no sample file for the main dump (`bioconcepts2pubtatorcentral.offset.gz`). I reached out to the folks at PubTator and they agreed to add a sample file for us (`bioconcepts2pubtatorcentral.offset.sample`). Should I go ahead and implement this data loader using that sample file for the time being? That would speed things up as I won't have to worry about loading from the API.

galtay commented:
> > hi @JohnGiorgi great work so far! can I ask that you implement this using the files that end in `sample` from the FTP for now so that we can try some quick local testing?
> 
> Hi @galtay, sorry I hadn't forgot about this PR! Just focused on working on two other data loaders first. When you originally suggested the above there was no sample file for the main dump (`bioconcepts2pubtatorcentral.offset.gz`). I reached out to the folks at PubTator and they added a sample file for us (`bioconcepts2pubtatorcentral.offset.sample`). Should I go ahead and implement this data loader using that sample file for the time being? That would speed things up as I won't have to worry about loading from the API.

oh great! thanks for reaching out to them and thanks to them for adding the sample file! Yes, lets use the sample file to avoid the complexities of the API if possible.

JohnGiorgi commented:
@galtay Okay this is ready for another look. Sorry for all the back and forth, this was the trickest of the three dataloader I took on by far! 

I incorporated most or all of the suggestions:

- The data loader loads a much smaller sample file in the same format as the full ~17GB dump.
- Added a module-level constant `_DB_NAME_TO_URL` which maps `db_name`'s to URLs. It's unused in the current data loader by at least this information is documented somewhere :shrug:

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,291 @@
+                            {"db_name": db_name, "db_id": entity.id}
+                        )
+
+                unified_example["entities"] = list(unified_entities.values())
```
I think our (confusing) way of handling offsets is causing some problems here ... examining the bigbio_kb output I see things like, 
```  {'id': '35157371_253959',
   'type': 'Gene',
   'text': ['TULIP-1', 'TULIP-1'],
   'offsets': [[3626, 3633], [4193, 4200]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'},
    {'db_name': 'ncbi_gene', 'db_id': '253959'}]},
```
the above should be two entity entries though, 
```
 {
   'id': '35157371_253959_a',
   'type': 'Gene',
   'text': ['TULIP-1'],
   'offsets': [[3626, 3633]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'}]
}, 
{
   'id': '35157371_253959_b',
   'type': 'Gene',
   'text': ['TULIP-1'],
   'offsets': [[4193, 4200]],
   'normalized': [{'db_name': 'ncbi_gene', 'db_id': '253959'}]
},
```

(same for all entries in `entities`) we have text as a list and offsets as nested list to handle the rare cases of single entity mentions that are represented by discontinuous spans ... but each entity mention (in the case of pubtator) will point to a single span ... we just have to represent the text as a list and the offsets as a nested list to be consistent.

JohnGiorgi replied:
Ah, I think I see, I was assuming entities with the same ID should get a single entry under `entities` and `text` and `offsets` should contain all of their unique mentions. I can fix this

galtay replied:
yes please!

galtay replied:
@sg-wbi just FYI

JohnGiorgi replied:
Fixed here, will have to fix in #20/#370 later!

JohnGiorgi replied:
Addressed here: https://github.com/bigscience-workshop/biomedical/pull/444

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,317 @@
+            schema="bigbio_kb",
+            subset_id="pubtator_central",
+        ),
+    ]
```
I didn't mean to discourage use of the full dataset here. I think it would be great to have both the `sample` dataset and the `full` dataset as options (so still four configs but now the API is replaced with the sample)

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,291 @@
+            description="PubTator Central BigBio schema",
+            schema="bigbio_kb",
+            subset_id="pubtator_central",
+        ),
```
```suggestion
        # sample source 
        BigBioConfig(
            name="pubtator_central_sample_source",
            version=SOURCE_VERSION,
            description="PubTator Central sample source schema",
            schema="source",
            subset_id="pubtator_central_sample",
        ),
        # sample big bio
        BigBioConfig(
            name="pubtator_central_sample_bigbio_kb",
            version=BIGBIO_VERSION,
            description="PubTator Central sample BigBio schema",
            schema="bigbio_kb",
            subset_id="pubtator_central_sample",
        ),
        # full dataset source
        BigBioConfig(
            name="pubtator_central_source",
            version=SOURCE_VERSION,
            description="PubTator Central source schema",
            schema="source",
            subset_id="pubtator_central",
        ),
        # full dataset bigbio
        BigBioConfig(
            name="pubtator_central_bigbio_kb",
            version=BIGBIO_VERSION,
            description="PubTator Central BigBio schema",
            schema="bigbio_kb",
            subset_id="pubtator_central",
        ),
```

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,322 @@
+
+    def _split_generators(self, dl_manager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+        urls = _URLS[_DATASETNAME]
```
here we will need to select the full dataset or sample dataset (probably using `self.config.subset_id`)

JohnGiorgi replied:
Done :+1:

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,322 @@
+including warranties of performance, merchantability or fitness for any particular purpose.
+"""
+
+_URLS = {_DATASETNAME: "https://ftp.ncbi.nlm.nih.gov/pub/lu/PubTatorCentral/bioconcepts2pubtatorcentral.offset.sample"}
```
will probably want one url pointing to the sample and one url pointing to the full dataset

galtay commented on biodatasets/pubtator_central/pubtator_central.py:
```diff
@@ -0,0 +1,322 @@
+            description="PubTator Central BigBio schema",
+            schema="bigbio_kb",
+            subset_id="pubtator_central",
+        ),
```
sorry for the confusion here, but i was trying to recommend 4 configs not 6

JohnGiorgi replied:
No worries, these are the four right? ({source, bigbio} * {sample, full}):

```python
    BUILDER_CONFIGS = [
        # sample source
        BigBioConfig(
            name="pubtator_central_sample_source",
            version=SOURCE_VERSION,
            description="PubTator Central sample source schema",
            schema="source",
            subset_id="pubtator_central_sample",
        ),
        # sample big bio
        BigBioConfig(
            name="pubtator_central_sample_bigbio_kb",
            version=BIGBIO_VERSION,
            description="PubTator Central sample BigBio schema",
            schema="bigbio_kb",
            subset_id="pubtator_central_sample",
        ),
        # full dataset source
        BigBioConfig(
            name="pubtator_central_source",
            version=SOURCE_VERSION,
            description="PubTator Central source schema",
            schema="source",
            subset_id="pubtator_central",
        ),
        # full dataset bigbio
        BigBioConfig(
            name="pubtator_central_bigbio_kb",
            version=BIGBIO_VERSION,
            description="PubTator Central BigBio schema",
            schema="bigbio_kb",
            subset_id="pubtator_central",
        ),
    ]
```

galtay replied:
yep

galtay commented:
the finish line is in sight! thanks so much for all the work on this one. I'll run it on the full dataset tonight to double check and if it looks good I'll merge

JohnGiorgi commented:
> the finish line is in sight! thanks so much for all the work on this one. I'll run it on the full dataset tonight to double check and if it looks good I'll merge

I, unfortunately, think there is a 100% chance it will fail, PubTator has all sorts of weird formatting issues (entity annotations that don't adhere to actually PubTator format) that are going to cause the `bioc` parser to fail :(

JohnGiorgi commented:
> > > the finish line is in sight! thanks so much for all the work on this one. I'll run it on the full dataset tonight to double check and if it looks good I'll merge
> > 
> > 
> > I, unfortunately, think it will fail at some point. PubTator has all sorts of weird edge cases -- entity annotations that don't actually adhere to PubTator format -- that are going to cause the `bioc` parser to fail :(
> 
> well its been running for about 12 hours now ;) so far, so good. I'll let it continue until it fails and if it does then we can figure out what to do,
> 
> ```
> python -m tests.test_bigbio biodatasets/pubtator_central/pubtator_central.py --subset_id pubtator_central
> INFO:__main__:args: Namespace(data_dir=None, path='biodatasets/pubtator_central/pubtator_central.py', schema=None, subset_id='pubtator_central', use_auth_token=None)
> INFO:__main__:self.PATH: biodatasets/pubtator_central/pubtator_central.py
> INFO:__main__:self.SUBSET_ID: pubtator_central
> INFO:__main__:self.SCHEMA: None
> INFO:__main__:self.DATA_DIR: None
> INFO:__main__:Checking for _SUPPORTED_TASKS ...
> INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.NAMED_ENTITY_RECOGNITION: 'NER'>, <Tasks.NAMED_ENTITY_DISAMBIGUATION: 'NED'>]
> INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'KB'}
> INFO:__main__:schemas_to_check: {'KB'}
> INFO:__main__:Checking load_dataset with config name pubtator_central_source
> Downloading and preparing dataset pubtator_central_dataset/pubtator_central_source to /Users/gabriel.altay/.cache/huggingface/datasets/pubtator_central_dataset/pubtator_central_source/2022.1.8/b5f88ddd7a4bc9bb63a4fcc8990f8f29f96f17bafe496ceccb6c625743fa04d2...
> Downloading data: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 18.5G/18.5G [19:49<00:00, 15.6MB/s]
> Dataset pubtator_central_dataset downloaded and prepared to /Users/gabriel.altay/.cache/huggingface/datasets/pubtator_central_dataset/pubtator_central_source/2022.1.8/b5f88ddd7a4bc9bb63a4fcc8990f8f29f96f17bafe496ceccb6c625743fa04d2. Subsequent calls will reuse this data.
> 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.50s/it]
> INFO:__main__:Checking load_dataset with config name pubtator_central_bigbio_kb
> Downloading and preparing dataset pubtator_central_dataset/pubtator_central_bigbio_kb to /Users/gabriel.altay/.cache/huggingface/datasets/pubtator_central_dataset/pubtator_central_bigbio_kb/1.0.0/b5f88ddd7a4bc9bb63a4fcc8990f8f29f96f17bafe496ceccb6c625743fa04d2...
> ```

Oh cool! Pleasantly surprised. Maybe the `bioc` parser is more resilient than I thought.

galtay commented:
this ran for about 36 hours on the full set but the sample set works fine so assuming the full set does too. going to merge now, but we can definitely revisit if trouble crops up with the full dataset ... also we should see about making the iterations a bit more efficient, but that can be phase 2 :)  excellent work @JohnGiorgi !

JohnGiorgi commented:
Awesome :) thanks for all your help! @galtay
